stan/math/opencl/double_d.hpp

```diff
@@ -0,0 +1,322 @@
+#ifndef STAN_MATH_OPENCL_DOUBLE_D_HPP
+#define STAN_MATH_OPENCL_DOUBLE_D_HPP
+
+#include <stan/math/opencl/stringify.hpp>
+
+#include <string>
+#include <cmath>
+#include <cfloat>
+#include <limits>
+
+/**
+ * Alternative stringify, that also exposes the code for use in C++ host code.
+ */
+#define STRINGIFY2(A) \
+#A;                 \
+  A
+namespace stan {
+namespace math {
+namespace internal {
+
+/**
+ * Double double - a 128 bit floating point number defined as an exact sum of 2
+ * doubles. `low` should be less than 1 ulp of `high`.
+ */
+typedef struct double_d {
+  double high;
+  double low;
+
+  double_d() {}
+  double_d(double a) {
+    high = a;
+    low = 0;
+  }
+  double_d(double h, double l) {
+    high = h;
+    low = l;
+  }
+} double_d;
+
+using std::copysign;
+using std::fabs;
+using std::isfinite;
+using std::isinf;
+using std::isnan;
+using std::ldexp;
+
+// \cond
+const char* double_d_src = STRINGIFY(
+    // \endcond
+    typedef struct {
+      double high;
+      double low;
+    } double_d;
+    // \cond
+    )
+    STRINGIFY2(
+        // \endcond
+
+        //    inline double_d mul_d_d(double a, double b) {
+        //      double_d res;
+        //      volatile double temp = a * b;
+        //      res.high = temp;
+        //      res.low = fma(a, b, -temp);
+        //      return res;
+        //    }
```

Comment:
This alternative implementation of multiplication of two doubles into double double using fma should be much faster, but for some reason I can't get it to work correctly. It might be compiler is optimizing the code too much. Does anyone have any ideas how to fix this?

Reply:
Missed this. This doesnt work with C++ or OpenCL or both?

Reply:
I think we can proceed with what is written currently. I can imagine you dug deep into this already and I dont have the time to dig into this right now so I would personally remove the commented code and optionally open an issue if someone wants to play with this at a later point. This PR looks great otherwise.

Reply:
I thoroughly tested this only in C++ (with the test that compares to quadruple precision). However we need this to work in both C++ and OpenCL. 

As for digging deep, I only checked that compiler does not do something funny and it actually generates the fma instruction (actually it is a call to library function, but I think that should be fine). Than I was out of ideas.

Reply:
Ok, if you want to take a few moments to look into this more, that is fine with me. But I am also fine with going with the current implementation. I would only remove the commented out code in that case.